state/src/storage/tx/matching_pools.rs

```diff
@@ -0,0 +1,206 @@
+
+// ---------------
+// | Key Helpers |
+// ---------------
+
+/// The key for the set of all matching pools
+pub fn all_pools_key() -> String {
+    "all-pools".to_string()
+}
+
+/// The key for the given order's matching pool
+pub fn matching_pool_key(order_id: &OrderIdentifier) -> String {
+    format!("{}{}", POOL_KEY_PREFIX, order_id)
+}
+
+// -----------
+// | Getters |
+// -----------
+
+impl<'db, T: TransactionKind> StateTxn<'db, T> {
+    /// Get the name of the matching pool the given order is in, if it's been
+    /// allocated to one
+    pub fn get_pool_for_order(
+        &self,
+        order_id: &OrderIdentifier,
+    ) -> Result<Option<String>, StorageError> {
+        let pool_key = matching_pool_key(order_id);
+        self.inner().read(POOL_TABLE, &pool_key)
+    }
+
+    /// Whether or not a pool with the given name exists
+    pub fn pool_exists(&self, pool_name: &String) -> Result<bool, StorageError> {
+        let all_pools_key = all_pools_key();
+        let all_pools = self.read_set(POOL_TABLE, &all_pools_key)?;
+        Ok(all_pools.contains(pool_name))
+    }
+
+    /// Whether or not the matching pool is empty
+    pub fn pool_is_empty(&self, pool_name: &String) -> Result<bool, StorageError> {
+        // We iterate over the mapping from orders -> their matching pool, and check
+        // if any orders are in the given pool
+        let cursor = self
+            .inner()
+            .cursor::<String, String>(POOL_TABLE)?
+            .with_key_filter(|key| key.starts_with(POOL_KEY_PREFIX));
+
+        let mut pool_in_use = false;
+        for pool_for_order in cursor.into_iter().values() {
+            if &pool_for_order? == pool_name {
+                pool_in_use = true;
+                break;
+            }
+        }
+
+        Ok(!pool_in_use)
+    }
+}
+
+// -----------
+// | Setters |
+// -----------
+
+impl<'db> StateTxn<'db, RW> {
+    /// Create a matching pool with the given name
+    pub fn create_matching_pool(&self, pool_name: &String) -> Result<(), StorageError> {
+        // Check that the pool does not already exist
+        if self.pool_exists(pool_name)? {
+            return Err(StorageError::Other(POOL_EXISTS_ERR.to_string()));
+        }
+
+        let all_pools_key = all_pools_key();
+        self.add_to_set(POOL_TABLE, &all_pools_key, pool_name)
+    }
+
+    /// Destroy a matching pool
+    pub fn destroy_matching_pool(&self, pool_name: &String) -> Result<(), StorageError> {
+        // Check that the pool is empty
+        if !self.pool_is_empty(pool_name)? {
+            return Err(StorageError::Other(POOL_NOT_EMPTY_ERR.to_string()));
+        }
+
+        let all_pools_key = all_pools_key();
+        self.remove_from_set(POOL_TABLE, &all_pools_key, pool_name)
+    }
+
+    /// Allocate an order to a matching pool.
+    ///
+    /// Note that we allow overwriting an order's pool, this signifies moving
+    /// the order from one pool to another.
+    pub fn allocate_order_to_pool(
```

Comment:
nit: i'd prefer `assign`; i don't think of "allocating" and object "to" another object, assigning is more accurate. Plus no order is "allocated"

Reply:
sure sgtm